samples/nrf9160/modem_shell/src/link/link_shell.c

```diff
@@ -1682,6 +1682,387 @@ int link_shell(const struct shell *shell, size_t argc, char **argv)
+
+	/* Reset getopt due to possible previous failures */
+	optreset = 1;
+	/* We start from subcmd arguments */
+	optind = 1;
+
+	int long_index = 0;
+	int opt;
+
+	while ((opt = getopt_long(argc, argv,
+				  "014",
+				  long_options, &long_index)) != -1) {
+		switch (opt) {
+		case '0':
+			funmode_option = LTE_LC_FUNC_MODE_POWER_OFF;
+			break;
+		case '1':
+			funmode_option = LTE_LC_FUNC_MODE_NORMAL;
+			break;
+		case '4':
+			funmode_option = LTE_LC_FUNC_MODE_OFFLINE;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_LTEOFF:
+			funmode_option = LTE_LC_FUNC_MODE_DEACTIVATE_LTE;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_LTEON:
+			funmode_option = LTE_LC_FUNC_MODE_ACTIVATE_LTE;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_GNSSOFF:
+			funmode_option = LTE_LC_FUNC_MODE_DEACTIVATE_GNSS;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_GNSSON:
+			funmode_option = LTE_LC_FUNC_MODE_ACTIVATE_GNSS;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_UICCOFF:
+			funmode_option = LTE_LC_FUNC_MODE_DEACTIVATE_UICC;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_UICCON:
+			funmode_option = LTE_LC_FUNC_MODE_ACTIVATE_UICC;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_FLIGHTMODE_UICCON:
+			funmode_option = LTE_LC_FUNC_MODE_OFFLINE_UICC_ON;
+			break;
+		case LINK_SHELL_OPT_NMODE_NO_REL14:
+			funmode_option = LTE_LC_FUNC_MODE_NORMAL;
+			nmode_use_rel14 = false;
+			break;
+		default:
+			break;
+		}
+	}
+
+	if (common_option == LINK_COMMON_READ) {
+		ret = lte_lc_func_mode_get(&functional_mode);
+		if (ret) {
+			mosh_error("Cannot get functional mode: %d", ret);
+		} else {
+			mosh_print(
+				"Functional mode read successfully: %s",
+				link_shell_funmode_to_string(functional_mode, snum));
+		}
+	} else if (funmode_option != LINK_FUNMODE_NONE) {
+		ret = link_func_mode_set(funmode_option, nmode_use_rel14);
+		if (ret < 0) {
+			mosh_error("Cannot set functional mode: %d", ret);
+		} else {
+			mosh_print(
+				"Functional mode set successfully: %s",
+				link_shell_funmode_to_string(
+					funmode_option, snum));
+		}
+	} else {
+		goto show_usage;
+	}
+	return ret;
+
+show_usage:
+	/* Reset getopt for another users: */
+	optreset = 1;
+
+	link_shell_print_usage(LINK_CMD_FUNMODE);
+	return ret;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	sub_link,
+	SHELL_CMD_ARG(
+		connect, NULL,
+		"Connect to given apn by creating and activating a new PDP context.",
+		link_connect, 0, 10),
+	SHELL_CMD_ARG(
+		disconnect,
+		NULL,
+		"Disconnect from given apn by deactivating and destroying a PDP context.",
+		link_disconnect, 0, 10),
+	SHELL_CMD_ARG(
+		status,
+		NULL,
+		"Show status of the current connection (no options).",
+		link_status, 0, 10),
+	SHELL_CMD_ARG(
+		funmode,
+		NULL,
+		"Set/read functional modes of the modem.",
+		link_funmode, 0, 10),
+
+	/* TODO: Remove this old parsing that is still kept to verify things */
+	SHELL_CMD(
+		old, NULL,
+		"Old parsing. Add subcommand and options after 'old' subcommand.",
+		link_shell),
```

Comment:
This is here just so you can use full functionality while refactoring is ongoing. Also useful for verification. E.g.,
link old sysmode -M

Reply:
Cleaned up old implementation now